packages/block-library/src/columns/edit.native.js

```diff
@@ -0,0 +1,279 @@
+ * In columns block, the only block we allow is 'core/column'.
+ *
+ * @constant
+ * @type {string[]}
+ */
+const ALLOWED_BLOCKS = [ 'core/column' ];
+
+/**
+ * Number of columns to assume for template in case the user opts to skip
+ * template option selection.
+ *
+ * @type {number}
+ */
+const DEFAULT_COLUMNS = 2;
+const MIN_COLUMNS_NUMBER = 1;
+
+const BREAKPOINTS = {
+	mobile: 480,
+	large: 768,
+};
+
+function ColumnsEditContainer( {
+	attributes,
+	updateAlignment,
+	updateColumns,
+	columnCount,
+	isSelected,
+	onAddNextColumn,
+	onDelete,
+} ) {
+	const [ resizeListener, sizes ] = useResizeObserver();
+	const [ columnsSettings, setColumnsSettings ] = useState( {
+		columnsInRow: MIN_COLUMNS_NUMBER,
+	} );
+
+	const { verticalAlignment } = attributes;
+	const { width } = sizes || {};
+
+	useEffect( () => {
+		const newColumnCount = ! columnCount ? DEFAULT_COLUMNS : columnCount;
+		updateColumns( columnCount, newColumnCount );
+		setColumnsSettings( {
+			...columnsSettings,
+			columnsInRow: getColumnsInRow( width, newColumnCount ),
+		} );
+	}, [ columnCount ] );
+
+	useEffect( () => {
+		setColumnsSettings( {
+			columnsInRow: getColumnsInRow( width, columnCount ),
+			width,
+		} );
+	}, [ width ] );
+
+	const getColumnsInRow = ( containerWidth, columnsNumber ) => {
+		if ( containerWidth < BREAKPOINTS.mobile ) {
+			// show only 1 Column in row for mobile breakpoint container width
+			return 1;
+		} else if ( containerWidth < BREAKPOINTS.large ) {
+			// show 2 Column in row for large breakpoint container width
+			return Math.min( columnCount, 2 );
+		}
+		// show all Column in one row
+		return columnsNumber;
+	};
+
+	const renderAppender = () => {
+		if ( isSelected ) {
+			return (
+				<InnerBlocks.ButtonBlockAppender
+					customOnAdd={ onAddNextColumn }
+				/>
+			);
+		}
+		return null;
+	};
+
+	return (
+		<>
+			<InspectorControls>
+				<PanelBody title={ __( 'Columns Settings' ) }>
+					<StepperControl
+						label={ __( 'Number of columns' ) }
+						icon="columns"
+						value={ columnCount }
+						onChange={ ( value ) =>
+							updateColumns( columnCount, value )
+						}
+						min={ MIN_COLUMNS_NUMBER }
+						max={ columnCount + 1 }
+						separatorType={ 'none' }
+					/>
+				</PanelBody>
+			</InspectorControls>
+			<BlockControls>
+				<BlockVerticalAlignmentToolbar
+					onChange={ updateAlignment }
+					value={ verticalAlignment }
+				/>
+			</BlockControls>
+			<View style={ isSelected && styles.innerBlocksSelected }>
+				{ resizeListener }
+				<InnerBlocks
+					renderAppender={ renderAppender }
+					__experimentalMoverDirection={
+						getColumnsInRow( width, columnCount ) > 1
+							? 'horizontal'
+							: undefined
+					}
+					flatListProps={ {
+						contentContainerStyle: styles.columnsContainer,
+						horizontal: true,
+						scrollEnabled: false,
+						style: styles.innerBlocks,
+					} }
+					allowedBlocks={ ALLOWED_BLOCKS }
+					customBlockProps={ {
```

Comment:
We have both flatListProps and other style props inside "customBlockProps"

For someone who don't know about the inner implementation of "InnerBlocks" this API is quite hard to grasp and use. And ideally we should encapsulate/hide FlatList usage, one shouldn't know we are using flatList inside and if one day we use a different kind of scrollable component inside we should only change the inner implementation instead of updating all InnerBlocks usages in the code base.

1. How can we make this simpler and more encapsulated in a way that consumer of this component won't even know we are using a FlatList inside?

2. Do we really need to pass contentContainerStyle from outside? Can we instead pass a well defined prop and apply this inside? Same for style: styles.innerBlocks. 

3. I believe we can pass "horizontal" and "scrollEnabled" as an InnerBlocks prop.

Reply:
BTW, if we already know that scrolling is done only on root level, why not use the "isRootList" inside /block-list/index.native.js this and get rid of scrollEnabled prop completely?

Reply:
>BTW, if we already know that scrolling is done only on root level, why not use the "isRootList" inside /block-list/index.native.js this and get rid of scrollEnabled prop completely?

I have also propose this solution [here](https://github.com/WordPress/gutenberg/pull/19013/files#r402206712) so it seems we are on the same page

>I believe we can pass "horizontal" and "scrollEnabled" as an InnerBlocks prop.

According to above we only should pass `horizontal` then

>Do we really need to pass contentContainerStyle from outside? Can we instead pass a well defined prop and apply this inside? Same for style: styles.innerBlocks.

contentContainerStyle - we can try to apply that style in `index.native.js` of BlockList after check if horizontal prop that will be passed from Columns is set to true (not sure if is should be applied to all block that in future might has horizontal set to true as well)

styles.innerBlocks - we can rename it to `styles.overflow` because it only change the overflow to visible to allow absolute positioning frame be visible

Reply:
> contentContainerStyle - we can try to apply that style in index.native.js of BlockList after check if horizontal prop that will be passed from Columns is set to true (not sure if is should be applied to all block that in future might has horizontal set to true as well)

It doesn't have to be decided _only_ by "horizontal" prop, plus, it might not be enough. Buttons might probably need a different contentContainerStyle. Could you collaborate with @lukewalczak to come up with a prop structure to represent different types of horizontal layouts instead of needing to pass the whole style definition from outside?

> styles.innerBlocks - we can rename it to styles.overflow because it only change the overflow to visible to allow absolute positioning frame be visible

In which cases we want the overflow be **not** visible ? Is it something that can be decided internally? (by again using "isRootList" maybe?)

Reply:
>Buttons might probably need a different contentContainerStyle. Could you collaborate with @lukewalczak to come up with a prop structure to represent different types of horizontal layouts instead of needing to pass the whole style definition from outside?

Sure I will discuss it with Luke. As far as I remember Buttons do not need contentContainer because Button do not have InnerBlock inside. That's the main architecture difference between Button and Column

>In which cases we want the overflow be not visible ? Is it something that can be decided internally? (by again using "isRootList" maybe?)

That's something that we need to confirm if we can set it at BlockList level because I'm not sure how it affects rest of the blocks. In case of Columns without that the absolutely positioning frame was cropped

Reply:
In my latest changes I have split `flatListGroup` and refactor to skip passing `contentContainerStyle` from columns.

Only `style` prop left to be handled - working on this to align api with Buttons PR

Reply:
I have also discussed passing `style` prop with Luke (because it was common to Columns and Buttons) and we prepare solution that should work for both without passing the `style` as you asked.

`scrollEnabled` prop will be handled separately according to [this comment](https://github.com/WordPress/gutenberg/pull/19013#discussion_r402334335)